### ISSUE TITLE
Automatically set up HPKP pin in Chromium

### ISSUE THREAD

**Issue by JeremyRand** [namecoin/ncdns-nsis#5]
Right now, we ask the user to manually add an HPKP pin to Chromium.  This has a few problems:

1. It will scare many users who aren't used to messing with Chromium settings.
2. If the user makes a mistake (e.g. failing to check the `PKP` checkbox), the user loses security.
3. The Chromium UI's HPKP pins [expire after 1000 days](https://www.imperialviolet.org/2011/02/17/hstsui.html).  This means that the user loses security 1000 days after installing ncdns.

We should be able to install an HPKP pin automatically, which doesn't expire.  I'll add details to this issue later on how exactly this can be done.

**Comment by JeremyRand** [comment]
From my Windows 7 VM, adding this entry to Chromium's TransportSecurity file seems to do the job:

~~~
"nprXwzm7mHINFnNah1Seo5SG0Pz9vW6dTXsYyvEC/PQ=": {
      "dynamic_spki_hashes": [ "sha256/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=" ],
      "dynamic_spki_hashes_expiry": 99999999999.9999,
      "expiry": 99999999999.9999,
      "mode": "force-https",
      "pkp_include_subdomains": true,
      "pkp_observed": 1.0000,
      "report-uri": "",
      "sts_include_subdomains": false,
      "sts_observed": 1.0000
   }
~~~

`"nprXwzm7mHINFnNah1Seo5SG0Pz9vW6dTXsYyvEC/PQ="` is a hash of `bit` (Chromium invents their own stupid hash function for this purpose, don't ask me why they do that).  `"sha256/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="` is base64 encoding of 256 zero bits.  (This would presumably be changed to the safer nothing-up-my-sleeve hash that Ryan suggested.)

Note that since this was pulled out of a Windows 7 VM, it should be assumed to be malware until someone's actually audited that it's correct.  In particular, we need to be able to reproduce the hash of `bit`.

**Comment by JeremyRand** [comment]
The hash function is:

~~~
base64(sha256(byte(len("bit")) cat "bit" cat byte(0)))
~~~

Note that the hash function would be more complex for a domain name that isn't a TLD (each label is encoded separately).

I have no idea why Chromium uses this weird hash function.

**Comment by hlandau** [comment]
Looks like the hash of how DNS wire-encodes a name. It's still a little strange they use that rather than the human-readable representation, but it's not a format they've made up.

**Comment by JeremyRand** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Progress has been made here; I have a Go tool that will set up the
HPKP pin: https://github.com/namecoin/ncdns/pull/21 .

So, what's needed is:

1. Get the aforementioned PR merged.
2. NSIS script needs to check to make sure Chromium isn't running; if
Chromium is running, NSIS script needs to ask the user to close it.
(I don't know whether a filesystem lock will be placed on the
TransportSecurity file by Chromium.)
3. NSIS script needs to run the aforementioned Go tool against the
TransportSecurity file.
4. NSIS script no longer needs to display the tutorial.


-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJZQwEDAAoJELPy0WV4bWVw2BwQAKbvb5NmQCP/A8Agc+47F+4w
58bJ5b4mAFrTt+n02+QF8ahMA6JU99arLQGRVKkKPGv//p1PQbbSmR1LZdySC3uL
1UUnP72TyFU69bVrq3imEfU+6Jvlc3OLsZQ+8aOyLmr7BmTFZpk+F8GaF1L1G5W7
rSZdAeDoPLi31NA+CJjbmf1IfPJDRab2XGr4oJnPDJaLT2x47C59zzqGJ2uSc0Zh
AY+kP8JTnByD0Fm5HB6A8S576ot1tB378UGGM+l1KtYnMmW1pttwM9GE5RHrtvht
nNxGsz90cQpA+FzUhPx9uiMMDTx379agUmnNKBxwlbouG6DMY4jYUa729yDoysPQ
2MRtEUXXUTEtdg6lDhJ/jBal6vKDgAyOZnaKg7PZ1HMPPPrtVd4OOLyLZlm5JV8a
IV0Y+Dd7yY519BvB6/zsqHzOkeipWz1C2bZSvn/318a7wt/Dff4Vz455qnxTaWqU
jUJ3C/I5dMLrUjNaV52SZxjtK0EkUJXSNhQPrKhofd6JPPcyATBVsTMYhrxEVDlS
NnuOsXkx/IecJywu+hwTyVURgozf+fKoaHzgd1jSuQl7E30XyZfCaeBvRIKtpxnW
RxdE46j8I9aRig87VRiRINtqnno2IkeUSJ3hNszrdUC+EvMyHXNR/B5dpg6VwHOQ
dUGoKnERVGD/xcohRuHv
=EQgx
-----END PGP SIGNATURE-----

**Comment by JeremyRand** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

JeremyRand:
> Progress has been made here; I have a Go tool that will set up the 
> HPKP pin: https://github.com/namecoin/ncdns/pull/21 .

That PR is now merged.  @hlandau, would you like to make the needed
changes to the NSIS script?

-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJZSQs/AAoJELPy0WV4bWVw+zgP/R7nGGLWM9lPlXIGqoQ1NyuE
dNXC/hHJzVY5VU185a27YTE0pUEr1loGUpGvae9nJ5TRKpwWeaWeDZ9nQ2K3dArR
ugEGLlgwmzrSxpm0m48LKeVsou7GM4fcFlKauQaUgaPwnorzNjCReKC2GHz3LFfh
qPCO0bD9vVVqQTzqW9+gPQHy7HQ13Xuj6S+3dDVgQ/3MkOMnS8BHid+mEZDjjwlp
pAeYR06ECP8ekEdnbnbakh8qJ/Ot2de0Etf8PAjgH+M44r8LekmESuPa4UV7YWz1
rCcHIhYAcitLnl4Y5/PH3td+Esb0f9waGHt81fLdgb7q9tXBmg9W0+srubDGYg6p
yvF1C1CFslPIedRX0KZJ70YNwvOLREqR69Oh87kmqm6JFQILQYxr306+xnptRZ9G
14P1+LKO6qypDLZcYFWeTO0zB3wN6qZGupYvPv++OFUDdtcXm5x0xSErVYMGCKYf
4KjlunjIQ0PoJxIKSdg+fAs68yaXq6uUPZtir3seFyh2GicWe2OIod6PwPSYlwyO
CFKk+bMmRUyPQmPSTOR03vKz1QMD5xnaSGTqr4hFYE2rGcWvMDtMNOb7MVtqEMro
9374nm71trVSnMhjpYko2wBau8xAho49p70QNlbSQWl7Xj9OaBLfOruMGeaeAfOv
xfFO/kO3HIjlie3ZGusY
=VyL9
-----END PGP SIGNATURE-----

**Comment by JeremyRand** [comment]
This was fixed by #11, closing.
